README.md
[![Last Release](https://img.shields.io/github/tag/styx0x6/gremlins.svg?logo=github&label=Release&colorB=34bf49)](https://github.com/styx0x6/gremlins)
[![Travis CI Build Status - master](https://img.shields.io/travis/styx0x6/gremlins/master.svg?logo=travis&label=master)](https://travis-ci.org/styx0x6/gremlins)
[![Python 3.4+](https://img.shields.io/badge/Python-3.4+-blue.svg?logo=python)](https://www.python.org/)
[![License AGPL-3.0+](https://img.shields.io/badge/License-AGPL--3.0+-blue.svg)](https://raw.githubusercontent.com/styx0x6/gremlins/master/LICENSE)
[![Gitter Chat](https://img.shields.io/badge/Gitter-styx0x6%2fGremlins%20%e2%86%92-grey.svg?logo=gitter&colorB=ed1965)](https://gitter.im/styx0x6/Gremlins)

French version here: [[README-FR](README-FR.md)]

# Gremlins

About
----

**Gremlins** is an open source tool made to help you to protect your privacy by keeping the little *gremlins* out of your stuff!

All started [here](http://seclists.org/fulldisclosure/2011/May/434) and with the "HADOPI" french law.

A lot of entities in the world, like the HADOPI agency in France by example, are spying your private and self-hosted
internet-fronted services, like file sharing, especially on BitTorrent, and others... The fundamentals rights that belong to
each one of us are obviously to be able to block any of these entities that would like to be too much curious on our
stuff, and also, prevent them to access our files and services based on our simple desire!

If you can prevent anyone to physicaly access your home, you must also be able to do the same on your internet-fronted
services.

The main purpose of this tool is to facilitate the blocking of these *gremlins* onto and from your BitTorrent services.

**How it works?**

The script build a blocking policy based on IPv4/IPv6 addresses from **iBlockList** and the **RIPE** database. Requests
are based on predefined keywords in order to don't block the whole internet... The collected IP addresses are correlated and formatted. Then, the script had the capability to
automatically set the blocking policy on many devices:

* Linux-based servers via **iptables**
* **Sophos UTM 9** firewall (v9.408+)
* **FreeboxOS** Download (v3+, APIv4)
* NetGate **pfSense** firewall _- NOT YET IMPLEMENTED_
* Deciso **OPNsense** firewall _- NOT YET IMPLEMENTED_

Also, the script can simply generate the blocking list in CSV format:

    <CIDR_IP_RANGE>,<VERSION>,<SOURCE>,<MATCHED_KEYWORD>,<NAME>

_**Warning:**_

This tool has been made to help people to prevent them from being monitored by external entities (governmental or not).
You can protect, with the help of this tool, your sharing services in the same way you
would did it with your home against rubbers.

The fact of protecting your infrastructure and services from being stalked by strangers, named here *gremlins*, do
**NOT** allow you to share protected and copyrighted content without any permission. Gremlins developpers and
contributors are not responsible of your acts and they do **NOT** encourage you to do anything wrong or illegal.

_**The script doesn't guarantee a full protection against gremlins, the list isn't guaranteed as exhaustive as
it's generated from requests based on keywords. Also, it's possible that the generated list contains false-positives but
it's better to be safe than sorry.**_

Stay in the right way, do legal things, protect your privacy against *gremlins* and may the force be with you.

Installation
----

You can download the latest tarball by clicking [here](https://github.com/styx0x6/gremlins/tarball/master) or the
latest zipball by clicking [here](https://github.com/styx0x6/gremlins/zipball/master).

Preferably, you can download Gremlins by cloning the [Git](https://github.com/styx0x6/gremlins) repository:

    git clone --depth 1 https://github.com/styx0x6/gremlins.git gremlins

Gremlins works out of the box with [Python](http://www.python.org/download/) version **3.4** and more on any platform.
Libraries and dependancies are embedded within the script for better convenience.

Getting Started
----

Global options:

`python gremlins.py [-v | -h] [-d] [-c] [-Dc] [-De] [-D4 | -D6] [-Dr, -Di] <command> [<command_options...>]`

![help](help.png)

* List the IP addresses in CSV format:

`python gremlins.py list [-h]`

Links
----

* Homepage: https://github.com/styx0x6/gremlins
* Download: [.tar.gz](https://github.com/styx0x6/gremlins/tarball/master) or [.zip](https://github.com/styx0x6/gremlins/zipball/master)
* Git repository: git://github.com/styx0x6/gremlins.git
* Issues tracker: https://github.com/styx0x6/gremlins/issues

Roadmap & Changelog
----

All details are here: [[CHANGELOG](CHANGELOG.md)]

Contributing
----

Feel free to submit *issues* and enhancement *pull requests*!

[[Bugs & Support](https://github.com/styx0x6/gremlins/issues)]  
[[How to contribute to a project on Github](https://gist.github.com/MarcDiethelm/7303312)] by Marc Diethelm

[![Travis CI Build Status - dev](https://img.shields.io/travis/styx0x6/gremlins/dev.svg?logo=travis&label=dev)](https://travis-ci.org/styx0x6/gremlins)

Third-Party Libraries
----

* **Shields.io** - Badges as a service.

    ![License: CC0](https://img.shields.io/badge/License-CC0-lightgrey.svg)  
    [http://shields.io/](http://shields.io/)  
    [https://github.com/badges/shields/](https://github.com/badges/shields/)

* **Simple Icons** - Icons as a service.

    ![License: CC0](https://img.shields.io/badge/License-CC0-lightgrey.svg)  
    [https://simpleicons.org/](https://simpleicons.org/)  
    [https://github.com/simple-icons/simple-icons/](https://github.com/simple-icons/simple-icons/)

Gremlins includes bundled packages and below are their associated licensing terms:

* **colorama-0.3.7** - The Colorama library located in `thirdparty/colorama/`.

    Copyright (C) 2013, Jonathan Hartley.  
    ![License: BSD 3-Clause](https://img.shields.io/badge/License-BSD%203--Clause-orange.svg)  
    [https://pypi.org/project/colorama/](https://pypi.org/project/colorama/)  
    [https://github.com/tartley/colorama/](https://github.com/tartley/colorama/)

* **paramiko-2.1.1** - The Paramiko library located in `thirdparty/paramiko/`.

    Copyright (C) 2003-2011, Robey Pointer.  
    Copyright (c) 2013-2018, Jeff Forcier.  
    ![License: LGPL](https://img.shields.io/badge/License-LGPL-blue.svg)  
    [http://www.paramiko.org/](http://www.paramiko.org/)  
    [https://github.com/paramiko/paramiko/](https://github.com/paramiko/paramiko/)

* **requests-2.13.0** - The Requests library located in `thirdparty/requests/`.

    Copyright (C) 2016, Kenneth Reitz.  
    ![License: Apache 2.0](https://img.shields.io/badge/License-Apache%202.0-yellowgreen.svg)  
    [http://docs.python-requests.org/](http://docs.python-requests.org/)  
    [https://github.com/requests/requests/](https://github.com/requests/requests/)


CHANGELOG.md
# Roadmap

## Features

* Implement **NetGate pfSense** support.
* Implement **Deciso OPNsense** support.
* Others ISP home box.
* FreeboxOS One/Delta (API v5?) support.
* Deluge/uTorrent IPBlock auto configuration support.

## Fixes & Improvments

* SSH with certificates support (implementation + helps to be reviewed).
* A better management of keywords to include, with easy customization.
* Upgrade third-party libraries.
* iptables: add an option for logging `-l, --log` and logging path `--log-path`.

# Changelog

## Version 1.0.0 (*Coming soon...*)

* Full RIPE database support. *TODO*
* IPv6 full support. *TODO*
* Implemented **iptables** support. *IN PROGRESS*
* Implemented **Sophos UTM 9** support. *TODO*
* Implemented **FreeboxOS Download** support. *TODO*
* Added a french version of the README file.
* Added Travis build notifications on gitter (https://gitter.im/styx0x6/Gremlins).
* Reviewed README file with more running details and fixes.

## Version 0.1.1 (2018-11-03)

* Removed Python 3.7 from Travis CI build validation.

## Version 0.1.0 (2018-11-03)

First release with standard features:

* Build for Python 3.4+ (v3.3 mandatory for ipaddress module, v3.4 mandatory for enum module and Paramiko...).
* PEP8 code style with docstring.
* Build under AGPL-3.0+ license.
* Core structure ready arguments parsing and next features.
* iBlockList support.
* Format and manage IPs as CIDR.
* Display IPs in CSV format.
* Manage objects into an internal memory database.
* Code ready for a full IPv4 and IPv6 support. IPv6 not yet fully supported.
* Properly manage stdout/stderr and colorization.
* Handling custom exceptions and error messages properly.
* Documentation with `-h, --help`.
* Setup Travis CI build validation.


static/settings.py
#!/usr/bin/env python

# Copyright (C) 2017  @styx0x6
#
# This file is part of Gremlins.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


# TODO link information details to those into __init__.py?
"""
Gremlins project settings
"""
PROJECT_NAME = "Gremlins"
PROJECT_TITLE = "gremlins"
DESCRIPTION = "Protect your privacy, protect from gremlins!"
DESCRIPTION_2 = "Keep gremlins out of your stuff!"
LICENSE = "GNU Affero General Public License (AGPL-3.0+)"
LICENSE_SHORT = "AGPL-3.0+"
COPYRIGHT = "Copyright (C) 2018 @styx0x6"

VERSION = "0.2.0-dev"
VERSION_STRING = "%s/%s" % (PROJECT_TITLE, VERSION)
VALID_PYTHON_VERSION = "3.4"

AUTHOR = "@styx0x6"
AUTHOR_PAGE = "https://github.com/styx0x6"
HOME_PAGE = "https://github.com/styx0x6/gremlins"
ISSUES_PAGE = "https://github.com/styx0x6/gremlins/issues"
GIT_REPOSITORY = "git://github.com/styx0x6/gremlins.git"

"""
Gremlins static settings
"""
CMD_HELP = 'help'
CMD_LIST = 'list'
CMD_IPTABLES = 'iptables'
CMD_FBXOS = 'fbxos'
CMD_UTM9 = 'utm9'
CMD_PFSENSE = 'pfsense'
CMD_OPNSENSE = 'opnsense'

DEFAULT_SSH_PORT = '22'
DEFAULT_FBX_HOST = 'mafreebox.freebox.fr'
DEFAULT_FBX_HTTPS_PORT = '443'
DEFAULT_UTM9_HTTPS_PORT = '4444'
DEFAULT_PFSENSE_HTTPS_PORT = '443'
DEFAULT_OPNSENSE_HTTPS_PORT = '443'

# TODO Needed: CACHE_DIR = "cache"

# iBlockList static settings
IBL_HTTP_FILEFORMAT = 'p2p'
IBL_HTTP_ARCHIVEFORMAT = 'gz'
IBL_HTTP_URL = "http://list.iblocklist.com/index.php?list=%s&fileformat=%s&archiveformat=%s" \
               % ("%s", IBL_HTTP_FILEFORMAT, IBL_HTTP_ARCHIVEFORMAT)
IBL_LIST_ENC = 'utf-8'
IBL_SEP = ":"

# RIPE static settings
RIPE_HTTP_RESP_TYPE = 'json'
RIPE_HTTP_REST_URL = "https://rest.db.ripe.net/search.%s?query-string=%s&flags=no-filtering" \
                     % (RIPE_HTTP_RESP_TYPE, "%s")
RIPE_JSON_NAME_LIST = ['netname', 'descr']
RIPE_JSON_INETNUM = 'inetnum'

# Information gathering settings
IBL_LISTS = ['bt_level1', 'bt_level2']
KEYWORDS_LIST = ["hadopi", "tmg", "trident mediguard", "trident mediaguard", "trident medi guard",
                 "trident media guard"]

# iptables commands
IPTABLES_DROP = ""
IPTABLES_SAVE = ""

"""
Gremlins outputs settings
"""
# Default parameters for outputs
DEFAULT_COLORIZE_OUTPUT = True  # by default, output is colorized
DEFAULT_SHOW_RUNNING_OUTPUT = True  # by default, running output messages are shown
DEFAULT_SHOW_DEBUG_INFO = False  # by default, debug information is not shown
DEFAULT_SHOW_RUNNING_ERROR = True  # by default, running error messages are shown

"""
Banner definition
"""
BANNER = """
%sGremlins%s %s%s%s - %s

Home page: %s
Report bugs at: %s
Author: %s, License: %s

""" % ("%s", "%s", "%s", VERSION, "%s", DESCRIPTION_2, HOME_PAGE, ISSUES_PAGE, AUTHOR, LICENSE_SHORT)

"""
Help definitions
"""
HELP = """
Usage: %s [-v | -h] [-d] [-c] [-Dc] [-De] [-D4 | -D6] [-Dr, -Di] <command> [<command_options...>]

Options:

  -v, --version                 Show version and exit
  -h, --help                    Show this help message

  -d, --debug                   Enable verbose outputs
  -c, --clean-output            Do not show running output, keep only some commands results
  -Dc, --disable-coloring       Disable terminal coloring
  -De, --disable-error          Disable error messages

  -D4, --disable-ipv4           Disable IPv4 (TO BE IMPLEMENTED)
  -D6, --disable-ipv6           Disable IPv6 (TO BE IMPLEMENTED)

  -Dr, --disable-ripe           Disable RIPE database searching requests
  -Di, --disable-iblocklist     Disable iBlockList download requests

Commands:

  list                          Show the CIDR IP addresses list
  iptables                      Push iptables blocking ruleset
  fbxos (NYI)                   Push FreeboxOS Download Blacklist blocking ruleset (v3+, APIv4)
  utm9                          Push Sophos UTM 9 firewall blocking ruleset (v9.408+)
  pfsense (NYI)                 Push NetGate pfSense firewall blocking ruleset
  opnsense (NYI)                Push Deciso OPNsense firewall blocking ruleset

  For further details about commands, use: %s <command> -h.

"""

HELP_LIST = """
Usage: %s %s [-h]

Show the CIDR IP addresses list. Print each line in CSV format:

  <CIDR_IP_RANGE>,<VERSION>,<SOURCE>,<MATCHED_KEYWORD>,<NAME>

    where <VERSION> is '4' or '6'
    where <SOURCE> is 'RIPE','bt_level1'...

Options:

  -h, --help                    Show this help message

"""

HELP_IPTABLES = """
Usage: %s %s [-h] [-l] [-H <host> [-P <port>] [-u <user> -p <password>] [-Ds]]

Push iptables blocking ruleset using CLI via SSH.

Options:

  -h, --help                    Show this help message
  -l, --list                    List only the iptables commands, will not push them
  -H, --host                    The remote server hostname or IP address
  -P, --port                    The remote server SSH port, '22' by default
  -u, --user                    The username used to connect and push commands on the remote server
  -p, --password                The user's password to connect on the remote server
  -Ds, --do-not-save            Will not save the pushed commands, thus configuration will not be persistent

"""

HELP_FBXOS = """
Usage: %s %s [-h] [-H <host> [-P <port>] -u <user> -p <password>]

/!\ NOT YET IMPLEMENTED /!\\

Push FreeboxOS Download Blacklist blocking ruleset using the REST API (Freebox Revolution v3+, APIv4).

Options:

  -h, --help                    Show this help message
  -H, --host                    The remote Freebox hostname or IP address, 'mafreebox.freebox.fr' by default
  -P, --port                    The remote Freebox HTTPS port, '443' by default
  -u, --user                    The authentication username to get the token challenge
  -p, --password                The user's password to get the token challenge

"""

HELP_UTM9 = """
Usage: %s %s [-h] [-H <host> [-P <port>] [-t <REST token ID>] [-u <user> -p <password>] [-Dl]]

Push Sophos UTM 9 firewall blocking ruleset using the REST API (v9.408+).

Options:

  -h, --help                    Show this help message
  -H, --host                    The remote Sophos UTM 9 firewall hostname or IP address
  -P, --port                    The remote Sophos UTM 9 firewall HTTPS port, '4444' by default
  -t, --token                   The REST authentication token ID, if do not use user/password credentials
  -u, --user                    The authentication username on the REST service, if do not use a token ID
  -p, --password                The user's password to be authenticated on the REST service with credentials
  -Dl, --do-not-log             Set the generated rules with no logging

"""

HELP_PFSENSE = """
Usage: %s %s [-h] [-H <host> [-P <port>] [-u <user> -p <password>] [-Dl]]

/!\ NOT YET IMPLEMENTED /!\\

Push pfSense firewall blocking ruleset using the REST API.

Options:

  -h, --help                    Show this help message
  -H, --host                    The remote pfSense firewall hostname or IP address
  -P, --port                    The remote pfSense firewall HTTPS port, '443' by default
  -u, --user                    The authentication username to get the token challenge
  -p, --password                The user's password to get the token challenge
  -Dl, --do-not-log             Set the generated rules with no logging

"""

HELP_OPNSENSE = """
Usage: %s %s [-h] [-H <host> [-P <port>] [-u <user> -p <password>] [-Dl]]

/!\ NOT YET IMPLEMENTED /!\\

Push OPNsense firewall blocking ruleset using the REST API.

Options:

  -h, --help                    Show this help message
  -H, --host                    The remote OPNsense firewall hostname or IP address
  -P, --port                    The remote OPNsense firewall HTTPS port, '443' by default
  -u, --user                    The authentication username to get the token challenge
  -p, --password                The user's password to get the token challenge
  -Dl, --do-not-log             Set the generated rules with no logging

"""
